Second Red Black Tree Program

brennandoubt opened the pull request:
#### Description of Change
I think this variation of the red black tree data structure has helpful additional documentation and naming for functions/variables to make it clearer how different parts of a red black tree's source code work. I started the program as a college homework assignment but all of the code and implementation methods are my own. I'm happy to update or fix anything with my code if needed!
<!--
Thank you for your Pull Request. Please provide a description above and review
the requirements below.

Contributors guide: https://github.com/TheAlgorithms/C-Plus-Plus/CONTRIBUTING.md
-->

#### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: <!-- Please add a one-line description for developers or pull request viewers -->

Panquesito7 reviewed (changes requested):
Hey there! Thank you for submitting a PR! 😄 Sadly, this does not meet our requirements. Check out our [repository standards](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md). Please read them **carefully** and **follow them**. **Let us know if you need any help here or on our [Discord server](https://discord.gg/c7MnfGFGa6).** 🙂


__
[![sema-logo](https://app.semasoftware.com/img/sema-tray-logo.gif)](https://semasoftware.com/gh) &nbsp;**Summary:** :hammer_and_wrench: This code needs a fix&nbsp; | &nbsp;**Tags:** Not maintainable

brennandoubt commented:
Thanks for the reply! I'm working on fixing the code right now. If it just said that I tried to merge my branch again, then I did that by mistake and am still working on the repo.

brennandoubt commented:
I updated my branch with the updated red black tree program!

Panquesito7 commented on data_structures/rb_tree2.cpp:
```diff
@@ -0,0 +1,1486 @@
+}
+
+
+int main(){
```
```suggestion
/**
 * @brief Main function
 * @returns 0 on exit
 */
int main() {
```


__
[![sema-logo](https://app.semasoftware.com/img/sema-tray-logo.gif)](https://semasoftware.com/gh) &nbsp;**Summary:** :hammer_and_wrench: This code needs a fix&nbsp; | &nbsp;**Tags:** Inelegant

Panquesito7 commented on data_structures/rb_tree2.cpp:
```diff
@@ -0,0 +1,1486 @@
+    
+    cout << "ALL TESTS PASSED!!" << endl;
+    return 0;
+};
```
Missing an endline. Without this endline, there could be issues with some IDEs and configurations.
```suggestion
}

```


__
[![sema-logo](https://app.semasoftware.com/img/sema-tray-logo.gif)](https://semasoftware.com/gh) &nbsp;**Summary:** :hammer_and_wrench: This code needs a fix&nbsp; | &nbsp;**Tags:** Not maintainable

brennandoubt commented:
Should I make sure all linter suggestions are added to my code? I have a couple of tests that have assert statements over 80 characters long, but I wasn't sure if it'd be preferred to just remove the longer test lines I have.

brennandoubt commented:
Hi, I posted a question on the pull request about if I need to apply all of the linter suggestions for my program, mainly for my longer code lines that hold larger data structures to test on. I wasn’t sure if I just need to remove my larger tests for my program in order to pass the linter code check or if it’s possible to leave them and apply linter’s other suggestions for cleaning up my code aside from the suggestions to shorten certain lines of code.

Is there a specific time when the pull request will be closed before any more activity occurs on it?

Thanks in advance!
Brennan

> Am 02.10.2022 um 02:17 schrieb github-actions[bot] ***@***.***>:
> 
> 
> This pull request has been automatically marked as abandoned because it has not had recent activity. It will be closed if no further activity occurs. Thank you for your contributions.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/TheAlgorithms/C-Plus-Plus/pull/1968#issuecomment-1264513994>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AVQX5SA73OLNEKYP752IJA3WBDIB5ANCNFSM55SJ2AAQ>.
> You are receiving this because you authored the thread.
>

Panquesito7 commented:
Hello. It'd be great if you could fix all the linter warnings/errors. These are to ensure your code matches our guidelines.
You can always reach out to us if you need any help! Feel free to join our [**Discord server**](https://the-algorithms.com/discord). 🙂

brennandoubt commented:
I just updated my branch with all the linter errors fixed. If there's anything else I need to do with my program, please let me know!

Panquesito7 commented on data_structures/rb_tree2.cpp:
```diff
@@ -0,0 +1,1490 @@
+ * 
+ */
+
+#include <cassert>
+#include <random>
+#include <iostream>
+#include <stdexcept>
+#include <string>
+#include <sstream>
+#include <vector>
```
Please add a one-line description of what the library/header is for (**see the example below**).

```c++
#include <iostream>    /// for IO operations
#include <cassert>    /// for assert
```

Panquesito7 commented on data_structures/rb_tree2.cpp:
```diff
@@ -0,0 +1,1490 @@
+/**
+ * @file rbtree2.cpp
```
No need to add the filename. It's best to let Doxygen automatically detect it, otherwise, you'd have to specify the whole directory as well.
```suggestion
 * @file
```

Panquesito7 commented on data_structures/rb_tree2.cpp:
```diff
@@ -0,0 +1,1490 @@
+/**
+ * @file rbtree2.cpp
+ * @author Brennan Allen (brennandoubt@gmail.com)
```
This is to add your GitHub profile as well. :)
```suggestion
 * @author [Brennan Allen](https://github.com/brennandoubt) (brennandoubt@gmail.com)
```

Panquesito7 commented on data_structures/rb_tree2.cpp:
```diff
@@ -0,0 +1,1490 @@
+#include <sstream>
+#include <vector>
+
+using namespace std;
```
This is discouraged. Please strictly use `std::` for each of its members.

Panquesito7 commented on data_structures/rb_tree2.cpp:
```diff
@@ -0,0 +1,1490 @@
+        };
+};
+
+void TestSimpleConstructor(){
```
The test functions should be documented as well. Something simple like this should work.

```c++
/**
 * @brief Self-test by using a simple constructor
 * @returns void
 */
```

This is so that Doxygen detects the functions as well.

Panquesito7 reviewed (changes requested):
Awesome work on your code, by the way! 🚀

brennandoubt commented:
Thank you! I think I forgot to update the master branch of my fork with my code last time, but I just pushed my updated code again (with the most recent suggestions you mentioned implemented).

Panquesito7 commented on data_structures/rb_tree2.cpp:
```diff
@@ -0,0 +1,1704 @@
+ * @version 0.1
+ * @date 2022-12-29
+ *
+ * @copyright Copyright (c) 2022
```
Not sure there's a need for the copyright field here.

Panquesito7 reviewed (approved):
Overall, this is amazing! Thank you so much. 🚀

tjgurwara99 commented on data_structures/rb_tree2.cpp:
```diff
@@ -0,0 +1,1703 @@
+                    } else {
+                        // setting parent and child pointers
+                        // for the node's one child
+                        ParentToMePtr(node_to_delete) =
+                        MeToOnlyChildPtr(node_to_delete);
+
+                        MeToOnlyChildPtr(node_to_delete)->parent =
+                        node_to_delete->parent;
+
+                        replacement_node = MeToOnlyChildPtr(node_to_delete);
+                    }
+                }
+                // Now check Rotate/Recolor conditions...
+                if (root != nullptr) {
+                    // a replacement node that's red or at the root
+                    // just needs to be recolored black
+                    if (replacement_node == root ||
+                    replacement_node->color == COLOR_RED) {
+                        replacement_node->color = COLOR_BLACK;
+                    } else if (node_to_delete->color == COLOR_BLACK &&
+                    (replacement_node->color == COLOR_BLACK ||
+                    replacement_node->treatAsNull == true)) {
+                        // deleted node was black and replacement node is
+                        // black/null, causes double-blackness problem in
+                        // tree that needs to be fixed
+                        replacement_node->color = COLOR_DOUBLE_BLACK;
+                        fix_double_black(replacement_node);
+                    }
+                }
+                // if node to delete is a leaf, disconnect it from its
+                // parent after treating it as null to fix any
+                // double-blackness issues in tree
+                if (node_to_delete->treatAsNull == true) {
+                    ParentToMePtr(node_to_delete) = nullptr;
+                }
+                // finally delete node's memory and decrement tree size
+                delete node_to_delete;
+                numItems--;
+            }
+        }
```
Are you sure you need this many `if` statements?

```cpp
            if (node_to_delete->left != nullptr &&
            node_to_delete->right != nullptr) {
                int ios_val = InOrderSuccessor(node_to_delete)->val;
                Remove(ios_val);
                node_to_delete->val = ios_val;
                return
            }
            ... // no need for this else block you can write this directly into the function without the else block
            // the same can be applied to your other `if else` blocks.
```

I suggest looking up "guard clauses" and understanding why they help with this messy `if else` block 😄

brennandoubt replied:
Oh ok, I'll look them up and work on applying them to my code. Thank you for the advice!

tjgurwara99 commented on data_structures/rb_tree2.cpp:
```diff
@@ -0,0 +1,1686 @@
+                    "Insert(): Node is already in tree"
+                    ", duplicate nodes not allowed.");
+            }
+            RBTNode *node_to_insert = new RBTNode();
```
Might wanna use `auto` here

brennandoubt replied:
Why here? Just curious, and thanks again for the notes on my code here!

tjgurwara99 replied:
I usually have two reasons for using auto instead of using the actual type:
- when the type is already known from the right hand side of the operand (in this case `new RBTNode()` is a good indicator)
- when the type is kind of irrelevant (there are certain cases of its abuse but I won't go into the details of it here).

tjgurwara99 commented on data_structures/rb_tree2.cpp:
```diff
@@ -0,0 +1,1686 @@
+        }
+    }
+    // getting numbers to remove, starting from end of vector
+    int idx = random_numbers.size() - 1;
```
This type conversion is wrong and can cause bigger issues

Panquesito7 reviewed (changes requested):
Strange, there seem to be lots of [`clang-tidy`](https://github.com/TheAlgorithms/C-Plus-Plus/actions/runs/4539194834/jobs/7998834904?pr=1968) errors.
Please fix those errors. Do let us know if you need any help with that. 🙂

brennandoubt commented:
Just updated the code to fix the linter errors

Panquesito7 reviewed (changes requested):
There seem to be [a few errors](https://github.com/TheAlgorithms/C-Plus-Plus/actions/runs/4840294430/jobs/8634070466?pr=1968) (mostly about pointers and the use of `NULL`). Please fix them.
Let us know if you need any help with that! Hop to our [Discord community](https://the-algorithms.com/discord). 🙂

brennandoubt commented:
Oh ok, sorry I didn't catch those. I'll work on fixing them!

Panquesito7 commented:
Hey there, just checking in :) Do you need any help?

brennandoubt commented:
Hi! I don't have any specific questions right now, but I noticed that many of the remaining errors involve defining additional operators and some C++ concepts I wasn't as familiar with (such as smart pointers). I've been researching how to properly add and use these in my code, and I made several changes already to fix some of the smaller errors like using a 'constexpr' instead of macros. Should I push these updates to the remote branch now, or is it better to wait until I finish updating my code to handle the rest of the errors?

Panquesito7 commented:
> Should I push these updates to the remote branch now, or is it better to wait until I finish updating my code to handle the rest of the errors?

As you wish. 🙂

realstealthninja commented on data_structures/rb_tree2.cpp:
```diff
@@ -0,0 +1,1749 @@
+ * @author [Brennan Allen](https://github.com/brennandoubt)
+ * (brennandoubt@gmail.com)
+ * @brief A Red-Black Tree (RBT) program with test cases.
+ * @version 0.1
```
please add a @description tag

realstealthninja commented on data_structures/rb_tree2.cpp:
```diff
@@ -0,0 +1,1749 @@
+using std::cout;
+using std::endl;
+using std::invalid_argument;
+using std::ostringstream;
```
wow this is quite a nice way to handle not wanting to std:: :heart:  good job!

realstealthninja commented on data_structures/rb_tree2.cpp:
```diff
@@ -0,0 +1,1749 @@
+
+    /**
+     * @brief returns size of red black tree
+     *
```
needs a @ return tag

realstealthninja commented on data_structures/rb_tree2.cpp:
```diff
@@ -0,0 +1,1749 @@
+     * @brief Checks if passed node has a red child.
+     *
+     * @param n
+     * @return true
```
you have to specify when it returns this value 
i.e
```suggestion
     * @return true if the node HAS a red child 
```

realstealthninja commented on data_structures/rb_tree2.cpp:
```diff
@@ -0,0 +1,1749 @@
+     * (left left, left right, right right, right left) to
+     * fix double-blackness issue in tree.
+     *
+     * *Note: Assumed that the parameters given will be correct for this
```
```suggestion
     * @note Assumed that the parameters given will be correct for this
```
doxygen has tags for this!

realstealthninja commented on data_structures/rb_tree2.cpp:
```diff
@@ -0,0 +1,1749 @@
+
+    // shift node colors in this rotated subtree to
+    // the right (left to parent to right) - after left rotate
+    void ShiftColorsRight(RBTNode* sibling) {
```
proper documentation please
follow
````
/**
 * @brief <short not about function>
 * @param <parametername> <why parameter is required>
 */
````

realstealthninja commented on data_structures/rb_tree2.cpp:
```diff
@@ -0,0 +1,1749 @@
+     *
+     * @param n the node to get the child pointer for
+     * @return RBTNode*& the pointer from parent 'n'
+     * to its one non-null child
```
needs a 
@ return tag

brennandoubt commented:
Thank you for the comments! I just pushed some updates on documentation and comments in my code.

realstealthninja commented on data_structures/rb_tree2.cpp:
```diff
@@ -96,7 +99,7 @@ class RedBlackTree {
     }
 
-    // TODO(brennandoubt): copy assignment operator
+    // @todo implement copy assignment operator
```
please do not leave to-dos without implementing them

realstealthninja commented on data_structures/rb_tree2.cpp:
```diff
@@ -905,7 +927,8 @@ class RedBlackTree {
-     * @param node
+     * @param node the top node where a postfix string
+     * is generated from it and its children
      * @return string
```
the returns need to be descriptive!

brennandoubt commented:
Should I be concerned about the failing Doxygen and Awesome CI workflows when pushing my code? I noticed it fails the check when running cmake for windows and fetching a repository for gh-pages.

realstealthninja commented:
> Should I be concerned about the failing Doxygen and Awesome CI workflows when pushing my code? I noticed it fails the check when running cmake for windows and fetching a repository for gh-pages.

possibly but we'll get to it once we get to it

realstealthninja commented on data_structures/rb_tree2.cpp:
```diff
@@ -0,0 +1,1728 @@
+/**
+ * @file
+ * @author [Brennan Allen](https://github.com/brennandoubt)
+ * (brennandoubt@gmail.com)
+ * @brief A Red-Black Tree (RBT) program with test cases.
+ * @version 0.1
+ * @date 2022-12-29
+ * 
+ * Copyright (no copyright)
+ * 
+ * @details A program simulating the Red Black Tree algorithm
+ * using helper functions with descriptive function names and annotations.
+ */
```
```suggestion
/**
 * @file
 * @brief A Red-Black Tree (RBT) program with test cases.
 * @details A program simulating the Red Black Tree algorithm
 * using helper functions with descriptive function names and annotations.
 *
 * @author [Brennan Allen](https://github.com/brennandoubt)
 */
```
I think we should fit this to look like all the other files also i dont think doxygen accepts 2 links as a singular hyperlink

realstealthninja commented on data_structures/rb_tree2.cpp:
```diff
@@ -0,0 +1,1728 @@
+using std::string;
+using std::vector;
+
+/// defining colors (const instead of defined macros)
```
```suggestion
// defining colors (const instead of defined macros)
```

realstealthninja commented on data_structures/rb_tree2.cpp:
```diff
@@ -0,0 +1,1728 @@
+ *
+ */
+struct RBTNode {
+    /// nodes are red by default
+    /// null/root/red's child nodes are recolored to black
+    uint16_t color = COLOR_RED;
+    /// has int value to represent node itself (default 0)
+    int val = 0;
+
+    /// has a parent and left/right child nodes
```
these doc comments should probably be either double slashes or moved into the @details section of the struct

realstealthninja commented on data_structures/rb_tree2.cpp:
```diff
@@ -0,0 +1,1728 @@
+     */
+    ~RedBlackTree() { PostDelete(root); }
+
+    // Helper for deconstructor - delete nodes in post-order
```
```suggestion
    /**
     * @brief Deletes nodes in post-order
     * @param node <documentation here>
     */
```

realstealthninja commented on data_structures/rb_tree2.cpp:
```diff
@@ -0,0 +1,1728 @@
+     * in the Red Black Tree.
+     *
+     * @param value the value to check for
+     * @return true if the tree contains a node with this value
+     * @return false otherwise
```
```suggestion
     * @return true if the tree CONTAINS a node with this value
     * @return false if the tree DOES NOT contain a node with this value
```

realstealthninja commented on data_structures/rb_tree2.cpp:
```diff
@@ -0,0 +1,1728 @@
+
+/**
+ * @brief self-test by calling constructor
+ * @returns void
```
```suggestion
 * @return void
```

realstealthninja commented on data_structures/rb_tree2.cpp:
```diff
@@ -0,0 +1,1728 @@
+
+/**
+ * @brief simple insert of a value into tree
+ * @returns void
```
```suggestion
 * @return void
```

realstealthninja commented on data_structures/rb_tree2.cpp:
```diff
@@ -0,0 +1,1728 @@
+
+/**
+ * @brief test inserting two values into tree
+ * @returns void
```
```suggestion
 * @return void
```

realstealthninja commented on data_structures/rb_tree2.cpp:
```diff
@@ -0,0 +1,1728 @@
+/**
+ * @brief test inserting three nodes into tree
+ * (and making rotations occur)
+ * @returns void
```
```suggestion
 * @return void
```

realstealthninja commented on data_structures/rb_tree2.cpp:
```diff
@@ -0,0 +1,1728 @@
+
+/**
+ * @brief test inserting four values into tree
+ * @returns void
```
```suggestion
 * @return void
```

realstealthninja commented on data_structures/rb_tree2.cpp:
```diff
@@ -0,0 +1,1728 @@
+
+/**
+ * @brief test inserting five nodes into tree
+ * @returns void
```
```suggestion
 * @return void
```

realstealthninja commented on data_structures/rb_tree2.cpp:
```diff
@@ -0,0 +1,1728 @@
+/**
+ * @brief testing if string functions
+ * return correct values
+ * @returns void
```
```suggestion
 * @return void
```

realstealthninja commented on data_structures/rb_tree2.cpp:
```diff
@@ -0,0 +1,1728 @@
+/**
+ * @brief additional testing with inserts into tree
+ * using different values
+ * @returns void
```
should be consistent with either return or returns

realstealthninja commented on data_structures/rb_tree2.cpp:
```diff
@@ -0,0 +1,1728 @@
+/**
+ * @brief main function where all test functions are
+ * called
+ * @returns void
```
```suggestion
 * @return 0 on successful exit
```

realstealthninja commented on data_structures/rb_tree2.cpp:
```diff
@@ -0,0 +1,1728 @@
+ * @returns void
+ */
+int main() {
+    TestSimpleConstructor();
```
i think moving all of these tests into a function with calls all of them would be nice
tests(); // runs self test implementations

realstealthninja commented on data_structures/rb_tree2.cpp:
```diff
@@ -0,0 +1,1650 @@
+    bool treatAsNull = false;
+};
+
+class RedBlackTree {
```
needs class documentation

realstealthninja commented on data_structures/rb_tree2.cpp:
```diff
@@ -0,0 +1,1650 @@
+
+    /**
+     * @brief Removes node from tree and re-balances as needed
+     * @note
```
```suggestion
     * @details
```

realstealthninja commented on data_structures/rb_tree2.cpp:
```diff
@@ -0,0 +1,1650 @@
+    /**
+     * @brief Fixes double-blackness issue in tree
+     * 
+     * @note
```
```suggestion
     * @details
```

realstealthninja commented on data_structures/rb_tree2.cpp:
```diff
@@ -0,0 +1,1650 @@
+    /**
+     * @brief Re-balances tree based on the double-black
+     * node and its sibling conditions
+     * @note 
```
i think you can replace the note tags with detail tags
notes are supposed to be short and a reminder to the reader of something about the function

realstealthninja commented on data_structures/rb_tree2.cpp:
```diff
@@ -0,0 +1,1650 @@
+}
+
+/**
+ * @brief Runs all tests
```
```suggestion
 * @brief self-test implementation
```

realstealthninja commented on data_structures/rb_tree2.cpp:
```diff
@@ -0,0 +1,1650 @@
+ * @return 0 with exit
+ */
+int main() {
+    tests();
```
```suggestion
    tests(); // run self-test implementations 
```

brennandoubt commented:
I changed the note tags to details except for the notes that describe assumptions for how functions will be used in the program.

realstealthninja reviewed (dismissed):
Looks great me thanks for the contribution we really appreciate all the help! ❤️

realstealthninja commented:
Now all you have to do is fix the errors being caused by awesome ci. Treat them like a review! one of those include using null_ptr instead of null. Let us know if you need help with it

brennandoubt commented:
Sorry for the late response, but I understand and will have time to work on fixing those errors over the rest of this week.

realstealthninja commented:
> Sorry for the late response, but I understand and will have time to work on fixing those errors over the rest of this week.

take your time! let us know if you need any help!

brennandoubt commented:
Should I be concerned about the gh-pages step failing in the workflow?

realstealthninja commented on data_structures/rb_tree2.cpp:
```diff
@@ -0,0 +1,1628 @@
+     * @param n the node pointed to by its parent
+     * @return RBTNode*& the parent's pointer to the node
+     */
+    RBTNode*& ParentToMePtr(RBTNode* n) {
```
could you  please follow snake case?
ie
this_is_a_name_in_snake_case - snake case
YouAreUsingCamelCase - camel case

the repo uses camel case

brennandoubt replied:
Yeah I'll update those, I wasn't sure if I should since the contributing guidelines only mentioned having the file name in snake case but I get it for consistency

brennandoubt commented:
> repo

Thanks! I thought so but wanted to make sure I wasn't missing something.

brennandoubt commented:
> > Should I be concerned about the gh-pages step failing in the workflow? I updated my fork and think I also fixed the issue causing the windows compile check to fail.
> 
> 
> 
> not at all gh-pages can only be built on the main repo

Thanks! I thought so but wanted to make sure I wasn't missing something

brennandoubt commented:
And if there's anything else left to do, please let me know!

brennandoubt commented:
Hi, just checking again if there's anything else left to do for the RBT algorithm

realstealthninja reviewed (approved):
Looks good to me!

brennandoubt commented:
Hi, is there anything else that needs to be done for my program?

brennandoubt commented:
Any updates?

brennandoubt commented:
Hi, is there any update? Or could I have a re-review from @Panquesito7 on my code?

brennandoubt commented:
Any update?
